### DISCUSSION TITLE
Identify build-time vs run-time on server

### DISCUSSION THREAD

**Discussion by markoleavy** [vercel/next.js#66832] [App Router] [answered]
Hello!
I've tried to search for that question here and on google with no luck.
Is there a way (an env variable?) to know if a server process is running at build time or at run time?
I'd like to use a cached cdn endpoint for my queries while building static pages, but then been able to switch the same queries to a non-cached server for on-request revalidation.
```js
const endpoint = process.env.IS_BUILD_TIME ? 'https://my-cached-server.com' : 'https://my-fresh-data-server.com'
```

**Comment by SouthLink** [comment]
You can do this through the environment variable configuration file

> In general only one .env.local file is needed. However, sometimes you might want to add some defaults for the development (next dev) or production (next start) environment.
> 
> Next.js allows you to set defaults in .env (all environments), .env.development (development environment), and .env.production (production environment).
> 
> .env.local always overrides the defaults set.

[Default Environment Variables](https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#default-environment-variables)

**Comment by markoleavy** [comment]
Hi @SouthLink, thank you for your answer.
I know about env variables, but I'm to getting how to make it useful for my aim, as adding a new variable for production env will make it available on both build-time and runtime.

**Reply by icyJoseph** [reply]
Have you search on the discussions? IIRC, there used to be `process.env.NEXT_PHASE === PHASE_PRODUCTION_BUILD` but also there's a process.env that Next.js sets when building... I can try to look it up for you, but this has been answered in discussions before, though, not for App Router, I think.

**Reply by icyJoseph** [reply, marked as answer]
yeah, NEXT_PHASE is still there, and the PHASE_PRODUCTION_BUILD you get from next/constants, though these are not very well documented now, they used to be...

Though I'd say, you can also try to inject your own env var in next.config.js, using the `config` as function option.

**Reply by markoleavy** [reply]
Thanks a lot @icyJoseph, that is exactly what I was looking for!

**Reply by theoephraim** [reply]
@icyJoseph - from what I can tell this is no longer accurate. Logging `process.env.NEXT_PHASE` both in my `next.config.mjs` and in a specific page are both showing `undefined`.

**Reply by icyJoseph** [reply]
As said, these are not very well documented now, and although, some test/e2e files use them:

- https://github.com/search?q=repo%3Avercel%2Fnext.js%20NEXT_PHASE&type=code

it might not be the best idea, unless something else is not possible... that being said...

In next.config the value for `NEXT_PHASE` is not yet filled into the `env` object, by Next.js, you have to use the function form of the config:

```js
/** @type {import('next').NextConfig} */
const nextConfig = (phase, { defaultConfig }) => {
  console.log("config", phase);
  return defaultConfig;
};

export default nextConfig;
```

See here, https://stackblitz.com/edit/nextjs-psvynq?file=next.config.mjs

And, as for a page, I can see it (stackblitz doesn't like 14.2.4 to build), locally:

```console
yarn run v1.22.19
$ next build
config phase-production-build
config phase-production-build
  ▲ Next.js 14.2.4

   Creating an optimized production build ...
config phase-production-build
config phase-production-build
config phase-production-build
 ✓ Compiled successfully
 ✓ Linting and checking validity of types
 ✓ Collecting page data
   Generating static pages (0/6)  [    ]
page phase-production-build

page phase-production-build
 ✓ Generating static pages (6/6)
 ✓ Collecting build traces
 ✓ Finalizing page optimization
```
